Recursive modules are saved as functor parameters in the environment

Since 6ffc859, recursive modules are stored as functor parameters in the typing environment ([here](https://github.com/ocaml/ocaml/blob/331b5e64a174f47da6b37fa47bf23acdf1625399/typing/typemod.ml#L1811)). Therefore, the error message when trying to alias them is wrong : 

```ocaml
module rec X1 : sig end =
  (struct module A = X2 end : sig module A = X2 end)
and X2 : sig end = struct end
```
returns 
```ocaml
Error: Functor arguments, such as X1, cannot be aliased
```

While fixing the error message is easy ("Functor arguments *and recursive modules*, such as X1, cannot be aliased"), I suspect that there might be other issues of functor-related restrictions that would incorrectly apply to recursive modules. Note that here, preventing the aliasing is correct and [in line with the manual](https://ocaml.org/manual/5.3/modulealias.html#start-section)


<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

--- a/testsuite/tests/typing-modules/aliases.ml
+++ b/testsuite/tests/typing-modules/aliases.ml
@@ -911,3 +911,29 @@ module M :
 module N : sig val s : string end
 val s : string = "hello"
 |}]
+
+
+(* Aliases of a recursive module (within the definition) are disallowed *)
+module rec X1 : sig end = struct
+  module Inner : sig module X = X2 end = struct end
+end
+and X2 : sig end = struct end
+[%%expect {|
+Line 2, characters 21-34:
+2 |   module Inner : sig module X = X2 end = struct end
+                         ^^^^^^^^^^^^^
+Error: Functor arguments and recursive modules (within the
+       recursive definition), such as "X2", cannot be aliased
+|}]
+
+
+(* Aliases of a recursive module (outside of the definition) are allowed *)
+module rec X1 : sig end = struct end
+and X2 : sig end = struct end
+
+module X3 : sig module Inner = X1 end = struct module Inner = X1 end
+[%%expect {|
+module rec X1 : sig end
+and X2 : sig end
+module X3 : sig module Inner = X1 end
+|}]

--- a/testsuite/tests/typing-modules/pr13185.ml
+++ b/testsuite/tests/typing-modules/pr13185.ml
@@ -11,5 +11,5 @@ module type S1 = sig end
 Line 2, characters 41-53:
 2 | module type S2 = functor (X : S1) -> sig module M = X end
                                              ^^^^^^^^^^^^
-Error: Functor arguments, such as "X", cannot be aliased
+Error: Functor arguments and recursive modules, such as "X", cannot be aliased
 |}]

--- a/testsuite/tests/typing-modules/pr13185.ml
+++ b/testsuite/tests/typing-modules/pr13185.ml
@@ -11,5 +11,6 @@ module type S1 = sig end
 Line 2, characters 41-53:
 2 | module type S2 = functor (X : S1) -> sig module M = X end
                                              ^^^^^^^^^^^^
-Error: Functor arguments and recursive modules, such as "X", cannot be aliased
+Error: Functor arguments and recursive modules (within the
+       recursive definition), such as "X", cannot be aliased
 |}]

--- a/testsuite/tests/typing-sigsubst/sigsubst.ml
+++ b/testsuite/tests/typing-sigsubst/sigsubst.ml
@@ -208,7 +208,7 @@ type (_, _) eq = Refl : ('a, 'a) eq
 Line 8, characters 4-16:
 8 |     module N = M
         ^^^^^^^^^^^^
-Error: Functor arguments, such as "M", cannot be aliased
+Error: Functor arguments and recursive modules, such as "M", cannot be aliased
 |}]
 
 (* Checking that the uses of M.t are rewritten regardless of how they

--- a/testsuite/tests/typing-sigsubst/sigsubst.ml
+++ b/testsuite/tests/typing-sigsubst/sigsubst.ml
@@ -208,7 +208,8 @@ type (_, _) eq = Refl : ('a, 'a) eq
 Line 8, characters 4-16:
 8 |     module N = M
         ^^^^^^^^^^^^
-Error: Functor arguments and recursive modules, such as "M", cannot be aliased
+Error: Functor arguments and recursive modules (within the
+       recursive definition), such as "M", cannot be aliased
 |}]
 
 (* Checking that the uses of M.t are rewritten regardless of how they

--- a/typing/env.ml
+++ b/typing/env.ml
@@ -151,7 +151,7 @@ type summary =
   | Env_class of summary * Ident.t * class_declaration
   | Env_cltype of summary * Ident.t * class_type_declaration
   | Env_open of summary * Path.t
-  | Env_functor_arg of summary * Ident.t
+  | Env_not_aliasable of summary * Ident.t
   | Env_constraints of summary * type_declaration Path.Map.t
   | Env_copy_types of summary
   | Env_persistent of summary * Ident.t
@@ -168,7 +168,7 @@ let map_summary f = function
   | Env_class (s, id, d) -> Env_class (f s, id, d)
   | Env_cltype (s, id, d) -> Env_cltype (f s, id, d)
   | Env_open (s, p) -> Env_open (f s, p)
-  | Env_functor_arg (s, id) -> Env_functor_arg (f s, id)
+  | Env_not_aliasable (s, id) -> Env_not_aliasable (f s, id)
   | Env_constraints (s, m) -> Env_constraints (f s, m)
   | Env_copy_types s -> Env_copy_types (f s)
   | Env_persistent (s, id) -> Env_persistent (f s, id)
@@ -517,7 +517,7 @@ type t = {
   modtypes: (modtype_data, modtype_data) IdTbl.t;
   classes: (class_data, class_data) IdTbl.t;
   cltypes: (cltype_data, cltype_data) IdTbl.t;
-  functor_args: unit Ident.tbl;
+  not_aliasable: unit Ident.tbl;
   summary: summary;
   local_constraints: type_declaration Path.Map.t;
   flags: int;
@@ -722,7 +722,7 @@ let empty = {
   classes = IdTbl.empty; cltypes = IdTbl.empty;
   summary = Env_empty; local_constraints = Path.Map.empty;
   flags = 0;
-  functor_args = Ident.empty;
+  not_aliasable = Ident.empty;
  }
 
 let in_signature b env =
@@ -1443,14 +1443,14 @@ let find_modtype_expansion_lazy path env =
 let find_modtype_expansion path env =
   Subst.Lazy.force_modtype (find_modtype_expansion_lazy path env)
 
-let rec is_functor_arg path env =
+let rec is_aliasable path env =
   match path with
     Pident id ->
-      begin try Ident.find_same id env.functor_args; true
-      with Not_found -> false
+      begin try Ident.find_same id env.not_aliasable; false
+      with Not_found -> true
       end
-  | Pdot (p, _) | Pextra_ty (p, _) -> is_functor_arg p env
-  | Papply _ -> true
+  | Pdot (p, _) | Pextra_ty (p, _) -> is_aliasable p env
+  | Papply _ -> false
 
 (* Copying types associated with values *)
 
@@ -2191,10 +2191,10 @@ let _ =
 
 (* Insertion of bindings by identifier *)
 
-let add_functor_arg id env =
+let mark_not_aliasable id env =
   {env with
-   functor_args = Ident.add id () env.functor_args;
-   summary = Env_functor_arg (env.summary, id)}
+   not_aliasable = Ident.add id () env.not_aliasable;
+   summary = Env_not_aliasable (env.summary, id)}
 
 let add_value ?check ?shape id desc env =
   let addr = value_declaration_address env id desc in
@@ -2210,11 +2210,15 @@ and add_extension ~check ?shape ~rebind id ext env =
   let shape = shape_or_leaf ext.ext_uid shape in
   store_extension ~check ~rebind id addr ext shape env
 
-and add_module_declaration ?(arg=false) ?shape ~check id presence md env =
+and add_module_declaration ?(noalias=false) ?shape ~check id presence md env =
   let check =
     if not check then
       None
-    else if arg && is_in_signature env then
+    else if noalias && is_in_signature env then
+      (* While recursive modules are also added with the noalias flag when
+         typing the recursive definitions, they are then added back without the
+         flag (to be aliased from the outside), and therefore could not throw
+         the warning, leaving only functor parameters *)
       Some (fun s -> Warnings.Unused_functor_parameter s)
     else
       Some (fun s -> Warnings.Unused_module s)
@@ -2223,7 +2227,7 @@ and add_module_declaration ?(arg=false) ?shape ~check id presence md env =
   let addr = module_declaration_address env id presence md in
   let shape = shape_or_leaf md.mdl_uid shape in
   let env = store_module ~check id addr presence md shape env in
-  if arg then add_functor_arg id env else env
+  if noalias then mark_not_aliasable id env else env
 
 and add_module_declaration_lazy ~update_summary id presence md env =
   let addr = module_declaration_address env id presence md in
@@ -2250,8 +2254,8 @@ and add_cltype ?shape id ty env =
   let shape = shape_or_leaf ty.clty_uid shape in
   store_cltype id ty shape env
 
-let add_module ?arg ?shape id presence mty env =
-  add_module_declaration ~check:false ?arg ?shape id presence (md mty) env
+let add_module ?noalias ?shape id presence mty env =
+  add_module_declaration ~check:false ?noalias ?shape id presence (md mty) env
 
 let add_module_lazy ~update_summary id presence mty env =
   let md = Subst.Lazy.{mdl_type = mty;
@@ -2289,9 +2293,9 @@ let enter_extension ~scope ~rebind name ext env =
   let env = store_extension ~check:true ~rebind id addr ext shape env in
   (id, env)
 
-let enter_module_declaration ~scope ?arg ?shape s presence md env =
+let enter_module_declaration ~scope ?noalias ?shape s presence md env =
   let id = Ident.create_scoped ~scope s in
-  (id, add_module_declaration ?arg ?shape ~check:true id presence md env)
+  (id, add_module_declaration ?noalias ?shape ~check:true id presence md env)
 
 let enter_modtype ~scope name mtd env =
   let id = Ident.create_scoped ~scope name in
@@ -2310,8 +2314,8 @@ let enter_cltype ~scope name desc env =
   let env = store_cltype id desc (Shape.leaf desc.clty_uid) env in
   (id, env)
 
-let enter_module ~scope ?arg s presence mty env =
-  enter_module_declaration ~scope ?arg s presence (md mty) env
+let enter_module ~scope ?noalias s presence mty env =
+  enter_module_declaration ~scope ?noalias s presence (md mty) env
 
 (* Insertion of all components of a signature *)
 
@@ -2454,7 +2458,7 @@ let remove_last_open root env0 =
     | Env_modtype _
     | Env_class _
     | Env_cltype _
-    | Env_functor_arg _
+    | Env_not_aliasable _
     | Env_constraints _
     | Env_persistent _
     | Env_copy_types _
@@ -3485,7 +3489,7 @@ let filter_non_loaded_persistent f env =
       | Env_class _
       | Env_cltype _
       | Env_open _
-      | Env_functor_arg _
+      | Env_not_aliasable _
       | Env_constraints _
       | Env_copy_types _
       | Env_persistent _

--- a/typing/env.mli
+++ b/typing/env.mli
@@ -41,7 +41,7 @@ type summary =
   | Env_open of summary * Path.t
   (** The string set argument of [Env_open] represents a list of module names
       to skip, i.e. that won't be imported in the toplevel namespace. *)
-  | Env_functor_arg of summary * Ident.t
+  | Env_not_aliasable of summary * Ident.t
   | Env_constraints of summary * type_declaration Path.Map.t
   | Env_copy_types of summary
   | Env_persistent of summary * Ident.t
@@ -116,8 +116,12 @@ val find_constructor_address: Path.t -> t -> address
 val shape_of_path:
   namespace:Shape.Sig_component_kind.t -> t -> Path.t -> Shape.t
 
-val add_functor_arg: Ident.t -> t -> t
-val is_functor_arg: Path.t -> t -> bool
+(** Indicates if a path [p] can be aliased, i.e. if
+    - it does not contain functor applications
+    - it does not start with a functor parameter
+    - it does not start with a recursive module (inside the recursive
+    definition) *)
+val is_aliasable: Path.t -> t -> bool
 
 val normalize_module_path: Location.t option -> t -> Path.t -> Path.t
 (* Normalize the path to a concrete module.
@@ -310,11 +314,11 @@ val add_type:
 val add_extension:
   check:bool -> ?shape:Shape.t -> rebind:bool -> Ident.t ->
   extension_constructor -> t -> t
-val add_module: ?arg:bool -> ?shape:Shape.t ->
+val add_module: ?noalias:bool -> ?shape:Shape.t ->
   Ident.t -> module_presence -> module_type -> t -> t
 val add_module_lazy: update_summary:bool ->
   Ident.t -> module_presence -> Subst.Lazy.modtype -> t -> t
-val add_module_declaration: ?arg:bool -> ?shape:Shape.t -> check:bool ->
+val add_module_declaration: ?noalias:bool -> ?shape:Shape.t -> check:bool ->
   Ident.t -> module_presence -> module_declaration -> t -> t
 val add_module_declaration_lazy: update_summary:bool ->
   Ident.t -> module_presence -> Subst.Lazy.module_decl -> t -> t
@@ -371,10 +375,10 @@ val enter_extension:
   scope:int -> rebind:bool -> string ->
   extension_constructor -> t -> Ident.t * t
 val enter_module:
-  scope:int -> ?arg:bool -> string -> module_presence ->
+  scope:int -> ?noalias:bool -> string -> module_presence ->
   module_type -> t -> Ident.t * t
 val enter_module_declaration:
-  scope:int -> ?arg:bool -> ?shape:Shape.t -> string -> module_presence ->
+  scope:int -> ?noalias:bool -> ?shape:Shape.t -> string -> module_presence ->
   module_declaration -> t -> Ident.t * t
 val enter_modtype:
   scope:int -> string -> modtype_declaration -> t -> Ident.t * t

--- a/typing/envaux.ml
+++ b/typing/envaux.ml
@@ -68,12 +68,12 @@ let rec env_from_summary sum subst =
           | Error `Functor -> assert false
           | Error `Not_found -> raise (Error (Module_not_found path'))
           end
-      | Env_functor_arg(Env_module(s, id, pres, desc), id')
+      | Env_not_aliasable(Env_module(s, id, pres, desc), id')
             when Ident.same id id' ->
           Env.add_module_declaration ~check:false
             id pres (Subst.module_declaration Keep subst desc)
-            ~arg:true (env_from_summary s subst)
-      | Env_functor_arg _ -> assert false
+            ~noalias:true (env_from_summary s subst)
+      | Env_not_aliasable _ -> assert false
       | Env_constraints(s, map) ->
           Path.Map.fold
             (fun path info ->

--- a/typing/includemod.ml
+++ b/typing/includemod.ml
@@ -511,7 +511,7 @@ let rec modtypes ~core ~direction ~loc env subst mty1 mty2 shape =
 and try_modtypes ~core ~direction ~loc env subst mty1 mty2 orig_shape =
   match mty1, mty2 with
   | (Mty_alias p1, Mty_alias p2) ->
-      if Env.is_functor_arg p2 env then
+      if not (Env.is_aliasable p2 env) then
         Error (Error.Invalid_module_alias p2)
       else if not (equal_module_paths env p1 subst p2) then
           Error Error.(Mt_core Incompatible_aliases)
@@ -1016,14 +1016,6 @@ and check_modtype_equiv ~core ~direction ~loc env mty1 mty2 =
 
 (* Simplified inclusion check between module types (for Env) *)
 
-let can_alias env path =
-  let rec no_apply = function
-    | Path.Pident _ -> true
-    | Path.Pdot(p, _) | Path.Pextra_ty (p, _) -> no_apply p
-    | Path.Papply _ -> false
-  in
-  no_apply path && not (Env.is_functor_arg path env)
-
 let core_inclusion = Core_inclusion.{
   type_declarations;
   value_descriptions;
@@ -1069,7 +1061,7 @@ exception Apply_error of {
   }
 
 let check_modtype_inclusion_raw ~loc env mty1 path1 mty2 =
-  let aliasable = can_alias env path1 in
+  let aliasable = Env.is_aliasable path1 env in
   let direction = Directionality.unknown ~mark:true in
   strengthened_modtypes ~core:core_inclusion ~direction ~loc ~aliasable env
     Subst.identity mty1 path1 mty2 Shape.dummy_mod
@@ -1088,7 +1080,7 @@ let check_functor_application_in_path
   | Error _errs ->
       if errors then
         let prepare_arg (arg_path, arg_mty) =
-          let aliasable = can_alias env arg_path in
+          let aliasable = Env.is_aliasable arg_path env in
           let smd = Mtype.strengthen ~aliasable env arg_mty arg_path in
           (Error.Named arg_path, smd)
         in
@@ -1278,7 +1270,7 @@ module Functor_app_diff = struct
            environment to track equalities with external components that the
            parameter might add. *)
         let mty = Subst.modtype Keep st.subst param_ty in
-        let env = Env.add_module ~arg:true param Mp_present mty st.env in
+        let env = Env.add_module ~noalias:true param Mp_present mty st.env in
         I.expand_params { st with env }
     | Keep ((Named arg,  _mty) , Named (Some param, _param), _) ->
         let res =
@@ -1294,7 +1286,7 @@ module Functor_app_diff = struct
     | Keep (((Anonymous|Empty_struct), mty),
             Named (Some param, _param), _) ->
         let mty' = Subst.modtype Keep st.subst mty in
-        let env = Env.add_module ~arg:true param Mp_present mty' st.env in
+        let env = Env.add_module ~noalias:true param Mp_present mty' st.env in
         let res = Option.map (Mtype.nondep_supertype env [param]) st.res in
         I.expand_params { st with env; res}
 

--- a/typing/mtype.ml
+++ b/typing/mtype.ml
@@ -214,7 +214,7 @@ let rec nondep_mty_with_presence env va ids pres mty =
       let res_env =
         match param with
         | None -> env
-        | Some param -> Env.add_module ~arg:true param Mp_present arg env
+        | Some param -> Env.add_module ~noalias:true param Mp_present arg env
       in
       let mty =
         Mty_functor(Named (param, nondep_mty env var_inv ids arg),

--- a/typing/out_type.ml
+++ b/typing/out_type.ml
@@ -1828,7 +1828,7 @@ and tree_of_functor_parameter = function
         | None -> None, fun env -> env
         | Some id ->
             Some (Ident.name id),
-            Env.add_module ~arg:true id Mp_present ty_arg
+            Env.add_module ~noalias:true id Mp_present ty_arg
       in
       Some (name, tree_of_modtype ~ellipsis:false ty_arg), env
 

--- a/typing/typemod.ml
+++ b/typing/typemod.ml
@@ -286,7 +286,7 @@ let iterator_with_env super env =
         match param with
         | None -> ()
         | Some id ->
-          env := lazy (Env.add_module ~arg:true id Mp_present
+          env := lazy (Env.add_module ~noalias:true id Mp_present
                        mty_arg (Lazy.force env_before))
       end;
       self.Btype.it_module_type self mty_body;
@@ -663,7 +663,7 @@ let merge_constraint_aux initial_env loc sg lid constr : merge_result =
     | Sig_module(id, _, md, _rs, _), [s], With_modsubst (lid',path,md')
       when Ident.name id = s ->
         let sig_env = Env.add_signature sg_for_env outer_sig_env in
-        let aliasable = not (Env.is_functor_arg path sig_env) in
+        let aliasable = Env.is_aliasable path sig_env in
         ignore
           (Includemod.strengthened_module_decl ~loc ~mark:true
              ~aliasable sig_env md' path md);
@@ -870,7 +870,7 @@ let rec approx_modtype env smty =
             let rarg = Mtype.scrape_for_functor_arg env arg in
             let scope = Ctype.create_scope () in
             let (id, newenv) =
-              Env.enter_module ~scope ~arg:true name Mp_present rarg env
+              Env.enter_module ~scope ~noalias:true name Mp_present rarg env
             in
             Types.Named (Some id, arg), newenv
       in
@@ -1404,8 +1404,8 @@ and transl_modtype_aux env smty =
                     md_uid = Uid.mk ~current_unit:(Env.get_current_unit ());
                   }
                 in
-                Env.enter_module_declaration ~scope ~arg:true name Mp_present
-                  arg_md env
+                Env.enter_module_declaration ~scope ~noalias:true
+                  name Mp_present arg_md env
               in
               Some id, newenv
           in
@@ -1551,7 +1551,7 @@ and transl_signature env sg =
             let pres =
               match tmty.mty_type with
               | Mty_alias p ->
-                  if Env.is_functor_arg p env then
+                  if not (Env.is_aliasable p env) then
                     raise (Error (pmd.pmd_loc, env, Cannot_alias p));
                   Mp_absent
               | _ -> Mp_present
@@ -1589,7 +1589,7 @@ and transl_signature env sg =
               Env.lookup_module ~loc:pms.pms_manifest.loc
                 pms.pms_manifest.txt env
             in
-            let aliasable = not (Env.is_functor_arg path env) in
+            let aliasable = Env.is_aliasable path env in
             let md =
               if not aliasable then
                 md
@@ -1808,7 +1808,7 @@ and transl_recmodule_modtypes env sdecls =
   let make_env curr =
     List.fold_left (fun env (id_shape, _, md, _) ->
       Option.fold ~none:env ~some:(fun (id, shape) ->
-        Env.add_module_declaration ~check:true ~shape ~arg:true
+        Env.add_module_declaration ~check:true ~shape ~noalias:true
           id Mp_present md env
       ) id_shape
     ) env curr
@@ -1924,7 +1924,7 @@ let rec nongen_modtype env = function
         | Unit
         | Named (None, _) -> env
         | Named (Some id, param) ->
-            Env.add_module ~arg:true id Mp_present param env
+            Env.add_module ~noalias:true id Mp_present param env
       in
       nongen_modtype env body
 
@@ -2054,7 +2054,8 @@ let check_recmodule_inclusion env bindings =
                  then mty_actual
                  else subst_and_strengthen env scope s (Some id) mty_actual
                in
-               Env.add_module ~arg:false ~shape id' Mp_present mty_actual' env)
+               Env.add_module ~noalias:false ~shape id'
+                 Mp_present mty_actual' env)
           env bindings1 in
       (* Build the output substitution Y_i <- X_i *)
       let s' =
@@ -2254,7 +2255,7 @@ and type_module_aux ~alias ~strengthen ~funct_body anchor env smod =
                  mod_env = env;
                  mod_attributes = smod.pmod_attributes;
                  mod_loc = smod.pmod_loc } in
-      let aliasable = not (Env.is_functor_arg path env) in
+      let aliasable = Env.is_aliasable path env in
       let shape =
         Env.shape_of_path ~namespace:Shape.Sig_component_kind.Module env path
       in
@@ -2321,7 +2322,7 @@ and type_module_aux ~alias ~strengthen ~funct_body anchor env smod =
               let id = Ident.create_scoped ~scope name in
               let shape = Shape.var md_uid id in
               let newenv = Env.add_module_declaration
-                ~shape ~arg:true ~check:true id Mp_present arg_md env
+                ~shape ~noalias:true ~check:true id Mp_present arg_md env
               in
               Some id, newenv, id
           in
@@ -2489,7 +2490,8 @@ and type_one_application ~ctx:(apply_loc,sfunct,md_f,args)
               | None -> env, mty_res
               | Some param ->
                   let env =
-                    Env.add_module ~arg:true param Mp_present arg.mod_type env
+                    Env.add_module ~noalias:true param
+                      Mp_present arg.mod_type env
                   in
                   check_well_formed_module env app_loc
                     "the signature of this functor application" mty_res;
@@ -3501,7 +3503,7 @@ let report_error ~loc _env = function
         (Style.as_inline_code path) p
   | Cannot_alias p ->
       Location.errorf ~loc
-        "Functor arguments, such as %a, cannot be aliased"
+        "Functor arguments and recursive modules, such as %a, cannot be aliased"
         (Style.as_inline_code path) p
   | Cannot_scrape_package_type p ->
       Location.errorf ~loc

--- a/typing/typemod.ml
+++ b/typing/typemod.ml
@@ -3503,7 +3503,8 @@ let report_error ~loc _env = function
         (Style.as_inline_code path) p
   | Cannot_alias p ->
       Location.errorf ~loc
-        "Functor arguments and recursive modules, such as %a, cannot be aliased"
+        "Functor arguments and@ recursive modules@ (within the@ recursive \
+         definition),@ such as %a,@ cannot be aliased"
         (Style.as_inline_code path) p
   | Cannot_scrape_package_type p ->
       Location.errorf ~loc

